{
 "cells": [
  {
   "cell_type": "code",
   "outputs": [],
   "source": [
    "with open(\"/home/davide/Downloads/KEGGCCMEE.html\") as f:\n",
    "    lines = f.readlines()\n",
    "\n",
    "#use beautifulsoup to parse the html\n",
    "from bs4 import BeautifulSoup\n",
    "soup = BeautifulSoup(open(\"/home/davide/Downloads/KEGGCCMEE.html\"), 'html.parser')\n",
    "#find all a tags and all b tags\n",
    "text=soup.get_text()\n",
    "text=text[:text.find(\"Organismal Systems\")]\n",
    "lines = text.split(\"\\n\")\n",
    "diz = {}\n",
    "for i,line in enumerate(lines):\n",
    "    if line.startswith(\"0\"):\n",
    "        if \"(\" in line:\n",
    "            current_key = line[:line.index(\"(\")].strip()\n",
    "        else:\n",
    "            current_key = line\n",
    "            #add to key lines[i+j] until lines[i+j] starts has (\n",
    "            j = 1\n",
    "            while not \" (\" in lines[i+j]:\n",
    "                current_key += lines[i+j]\n",
    "                j += 1\n",
    "        diz[current_key] = []\n",
    "\n",
    "    elif line.startswith(\"WP_\"):\n",
    "        proteins = line.split(\", \") if \", \" in line else [line]\n",
    "        diz[current_key].extend(proteins)\n",
    "#now make a dataframe that for every value of the dictionary has a row with the key as index and the keys of the dictionary in which appears as columns\n",
    "import pandas as pd\n",
    "\n",
    "# Get a list of all unique protein IDs\n",
    "all_proteins = set(protein for proteins in diz.values() for protein in proteins)\n",
    "\n",
    "# Initialize an empty list to hold the dataframes\n",
    "dfs = []\n",
    "\n",
    "# Iterate over each protein in the set of all proteins\n",
    "for protein in all_proteins:\n",
    "    # Create a dictionary where the keys are the pathways and the values are 1 if the protein is present in the pathway and 0 otherwise\n",
    "    data = {pathway: 1 if protein in proteins else 0 for pathway, proteins in diz.items()}\n",
    "    # Convert the dictionary into a DataFrame and append it to the list\n",
    "    df = pd.DataFrame(data, index=[protein])\n",
    "    dfs.append(df)\n",
    "\n",
    "# Concatenate all the dataframes into a single dataframe\n",
    "result = pd.concat(dfs)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-06T17:58:06.177030Z",
     "start_time": "2024-04-06T17:58:04.895001Z"
    }
   },
   "id": "d29d73ae11174b4e",
   "execution_count": 4
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "                01100 Metabolic pathways  \\\n                                       1   \nWP_250122003.1                         0   \nWP_250121290.1                         0   \nWP_250126236.1                         0   \nWP_250123796.1                         0   \n...                                  ...   \nWP_250121169.1                         1   \nWP_250126067.1                         1   \nWP_250126377.1                         1   \nWP_250123384.1                         1   \nWP_250123388.1                         1   \n\n                01110 Biosynthesis of secondary metabolites  \\\n                                                          0   \nWP_250122003.1                                            0   \nWP_250121290.1                                            0   \nWP_250126236.1                                            1   \nWP_250123796.1                                            0   \n...                                                     ...   \nWP_250121169.1                                            0   \nWP_250126067.1                                            1   \nWP_250126377.1                                            0   \nWP_250123384.1                                            1   \nWP_250123388.1                                            1   \n\n                01120 Microbial metabolism in diverse environments  \\\n                                                                0    \nWP_250122003.1                                                  0    \nWP_250121290.1                                                  0    \nWP_250126236.1                                                  0    \nWP_250123796.1                                                  0    \n...                                                           ...    \nWP_250121169.1                                                  0    \nWP_250126067.1                                                  0    \nWP_250126377.1                                                  1    \nWP_250123384.1                                                  0    \nWP_250123388.1                                                  1    \n\n                01200 Carbon metabolism  \\\n                                      0   \nWP_250122003.1                        0   \nWP_250121290.1                        0   \nWP_250126236.1                        0   \nWP_250123796.1                        0   \n...                                 ...   \nWP_250121169.1                        0   \nWP_250126067.1                        0   \nWP_250126377.1                        0   \nWP_250123384.1                        0   \nWP_250123388.1                        1   \n\n                01210 2-Oxocarboxylic acid metabolism  \\\n                                                    0   \nWP_250122003.1                                      0   \nWP_250121290.1                                      0   \nWP_250126236.1                                      0   \nWP_250123796.1                                      0   \n...                                               ...   \nWP_250121169.1                                      0   \nWP_250126067.1                                      0   \nWP_250126377.1                                      0   \nWP_250123384.1                                      0   \nWP_250123388.1                                      0   \n\n                01212 Fatty acid metabolism  \\\n                                          0   \nWP_250122003.1                            0   \nWP_250121290.1                            0   \nWP_250126236.1                            0   \nWP_250123796.1                            0   \n...                                     ...   \nWP_250121169.1                            0   \nWP_250126067.1                            1   \nWP_250126377.1                            0   \nWP_250123384.1                            0   \nWP_250123388.1                            0   \n\n                01230 Biosynthesis of amino acids  \\\n                                                0   \nWP_250122003.1                                  0   \nWP_250121290.1                                  0   \nWP_250126236.1                                  0   \nWP_250123796.1                                  0   \n...                                           ...   \nWP_250121169.1                                  0   \nWP_250126067.1                                  0   \nWP_250126377.1                                  0   \nWP_250123384.1                                  0   \nWP_250123388.1                                  0   \n\n                01232 Nucleotide metabolism  \\\n                                          0   \nWP_250122003.1                            0   \nWP_250121290.1                            0   \nWP_250126236.1                            0   \nWP_250123796.1                            0   \n...                                     ...   \nWP_250121169.1                            0   \nWP_250126067.1                            0   \nWP_250126377.1                            0   \nWP_250123384.1                            0   \nWP_250123388.1                            0   \n\n                01250 Biosynthesis of nucleotide sugars  \\\n                                                      0   \nWP_250122003.1                                        0   \nWP_250121290.1                                        0   \nWP_250126236.1                                        0   \nWP_250123796.1                                        0   \n...                                                 ...   \nWP_250121169.1                                        0   \nWP_250126067.1                                        0   \nWP_250126377.1                                        0   \nWP_250123384.1                                        0   \nWP_250123388.1                                        0   \n\n                01240 Biosynthesis of cofactors  ...  04113 Meiosis - yeast  \\\n                                              0  ...                      0   \nWP_250122003.1                                0  ...                      0   \nWP_250121290.1                                0  ...                      0   \nWP_250126236.1                                0  ...                      0   \nWP_250123796.1                                0  ...                      0   \n...                                         ...  ...                    ...   \nWP_250121169.1                                0  ...                      0   \nWP_250126067.1                                1  ...                      0   \nWP_250126377.1                                0  ...                      0   \nWP_250123384.1                                0  ...                      0   \nWP_250123388.1                                0  ...                      0   \n\n                04214 Apoptosis - fly  04216 Ferroptosis  04217 Necroptosis  \\\n                                    0                  0                  0   \nWP_250122003.1                      0                  0                  0   \nWP_250121290.1                      0                  0                  0   \nWP_250126236.1                      0                  0                  0   \nWP_250123796.1                      0                  0                  0   \n...                               ...                ...                ...   \nWP_250121169.1                      0                  0                  0   \nWP_250126067.1                      0                  0                  0   \nWP_250126377.1                      0                  0                  0   \nWP_250123384.1                      0                  0                  0   \nWP_250123388.1                      0                  0                  0   \n\n                02024 Quorum sensing  \\\n                                   0   \nWP_250122003.1                     0   \nWP_250121290.1                     0   \nWP_250126236.1                     0   \nWP_250123796.1                     0   \n...                              ...   \nWP_250121169.1                     0   \nWP_250126067.1                     0   \nWP_250126377.1                     0   \nWP_250123384.1                     0   \nWP_250123388.1                     0   \n\n                05111 Biofilm formation - Vibrio cholerae  \\\n                                                        0   \nWP_250122003.1                                          0   \nWP_250121290.1                                          0   \nWP_250126236.1                                          0   \nWP_250123796.1                                          0   \n...                                                   ...   \nWP_250121169.1                                          0   \nWP_250126067.1                                          0   \nWP_250126377.1                                          0   \nWP_250123384.1                                          0   \nWP_250123388.1                                          0   \n\n                02025 Biofilm formation - Pseudomonas aeruginosa  \\\n                                                               0   \nWP_250122003.1                                                 0   \nWP_250121290.1                                                 0   \nWP_250126236.1                                                 0   \nWP_250123796.1                                                 0   \n...                                                          ...   \nWP_250121169.1                                                 0   \nWP_250126067.1                                                 0   \nWP_250126377.1                                                 0   \nWP_250123384.1                                                 0   \nWP_250123388.1                                                 0   \n\n                02026 Biofilm formation - Escherichia coli  \\\n                                                         0   \nWP_250122003.1                                           0   \nWP_250121290.1                                           0   \nWP_250126236.1                                           0   \nWP_250123796.1                                           0   \n...                                                    ...   \nWP_250121169.1                                           0   \nWP_250126067.1                                           0   \nWP_250126377.1                                           0   \nWP_250123384.1                                           0   \nWP_250123388.1                                           0   \n\n                02030 Bacterial chemotaxis  02040 Flagellar assembly  \n                                         0                         0  \nWP_250122003.1                           0                         0  \nWP_250121290.1                           0                         0  \nWP_250126236.1                           0                         0  \nWP_250123796.1                           0                         0  \n...                                    ...                       ...  \nWP_250121169.1                           0                         0  \nWP_250126067.1                           0                         0  \nWP_250126377.1                           0                         0  \nWP_250123384.1                           0                         0  \nWP_250123388.1                           0                         0  \n\n[1207 rows x 172 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>01100 Metabolic pathways</th>\n      <th>01110 Biosynthesis of secondary metabolites</th>\n      <th>01120 Microbial metabolism in diverse environments</th>\n      <th>01200 Carbon metabolism</th>\n      <th>01210 2-Oxocarboxylic acid metabolism</th>\n      <th>01212 Fatty acid metabolism</th>\n      <th>01230 Biosynthesis of amino acids</th>\n      <th>01232 Nucleotide metabolism</th>\n      <th>01250 Biosynthesis of nucleotide sugars</th>\n      <th>01240 Biosynthesis of cofactors</th>\n      <th>...</th>\n      <th>04113 Meiosis - yeast</th>\n      <th>04214 Apoptosis - fly</th>\n      <th>04216 Ferroptosis</th>\n      <th>04217 Necroptosis</th>\n      <th>02024 Quorum sensing</th>\n      <th>05111 Biofilm formation - Vibrio cholerae</th>\n      <th>02025 Biofilm formation - Pseudomonas aeruginosa</th>\n      <th>02026 Biofilm formation - Escherichia coli</th>\n      <th>02030 Bacterial chemotaxis</th>\n      <th>02040 Flagellar assembly</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th></th>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>WP_250122003.1</th>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>WP_250121290.1</th>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>WP_250126236.1</th>\n      <td>0</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>WP_250123796.1</th>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>WP_250121169.1</th>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>WP_250126067.1</th>\n      <td>1</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>1</td>\n      <td>...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>WP_250126377.1</th>\n      <td>1</td>\n      <td>0</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>WP_250123384.1</th>\n      <td>1</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>WP_250123388.1</th>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>...</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n  </tbody>\n</table>\n<p>1207 rows × 172 columns</p>\n</div>"
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-06T17:58:06.887939Z",
     "start_time": "2024-04-06T17:58:06.876176Z"
    }
   },
   "id": "c99610a7a5d3dad1",
   "execution_count": 5
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "Index(['01100 Metabolic pathways',\n       '01110 Biosynthesis of secondary metabolites',\n       '01120 Microbial metabolism in diverse environments',\n       '01200 Carbon metabolism', '01210 2-Oxocarboxylic acid metabolism',\n       '01212 Fatty acid metabolism', '01230 Biosynthesis of amino acids',\n       '01232 Nucleotide metabolism',\n       '01250 Biosynthesis of nucleotide sugars',\n       '01240 Biosynthesis of cofactors',\n       ...\n       '04113 Meiosis - yeast', '04214 Apoptosis - fly', '04216 Ferroptosis',\n       '04217 Necroptosis', '02024 Quorum sensing',\n       '05111 Biofilm formation - Vibrio cholerae',\n       '02025 Biofilm formation - Pseudomonas aeruginosa',\n       '02026 Biofilm formation - Escherichia coli',\n       '02030 Bacterial chemotaxis', '02040 Flagellar assembly'],\n      dtype='object', length=172)"
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result.columns"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-06T17:58:07.867974Z",
     "start_time": "2024-04-06T17:58:07.861184Z"
    }
   },
   "id": "a2306e7b6fcc5ae6",
   "execution_count": 6
  },
  {
   "cell_type": "code",
   "outputs": [],
   "source": [
    "fimo_best=\"CCMEE29best_site.narrowPeak\"\n",
    "genbank_dir = \"annotati_Refseq\"\n",
    "import os\n",
    "import re\n",
    "from Bio import SeqIO\n",
    "from collections import Counter\n",
    "from scipy.stats import hypergeom\n",
    "import pandas as pd\n",
    "diz={\n",
    "\n",
    "    'LAU37' : 'Chroococcidiopsis_sp._CCMEE_29_(cyanobacteria)_GCF_023558375.gbff' ,\n",
    "\n",
    "}\n",
    "with open(fimo_best,\"r\") as f:\n",
    "    content=f.readlines()\n",
    "names = [line.split()[0] for line in content]\n",
    "kos = []\n",
    "kos_background = []\n",
    "ko_to_genes = {}\n",
    "i,j=0,0\n",
    "for file2 in os.listdir(genbank_dir):\n",
    "    genoma = SeqIO.parse(os.path.join(genbank_dir,file2), \"genbank\")\n",
    "    for seq in genoma:\n",
    "        for record in seq.features:\n",
    "            if record.type != \"CDS\":\n",
    "                continue\n",
    "            gene=record.qualifiers.get(\"gene\", [\"None\"])[0]\n",
    "            pid=record.qualifiers.get(\"protein_id\", [\"None\"])[0]\n",
    "            if record.qualifiers[\"locus_tag\"][0] in names:\n",
    "                i+=1\n",
    "                #see row of result with the protein id as index\n",
    "                if pid in result.index:\n",
    "                    #make a list of columns of the row with value 1\n",
    "                    ko_terms = result.columns[result.loc[pid] == 1].tolist()\n",
    "                    for ko_term in ko_terms:\n",
    "                        kos.append(ko_term)\n",
    "                        if ko_term in ko_to_genes:\n",
    "                            ko_to_genes[ko_term].append(gene)\n",
    "                        else:\n",
    "                            ko_to_genes[ko_term] = [gene]\n",
    "            else:\n",
    "                if pid in result.index:\n",
    "                    j+=1\n",
    "                    kos_background.extend(result.columns[result.loc[pid] == 1].tolist())\n",
    "kos = Counter(kos)\n",
    "kos_background = Counter(kos_background)\n",
    "M = i+j\n",
    "n = i\n",
    "p_values = {}\n",
    "enrichments = {}\n",
    "for ko in kos:\n",
    "    k = kos[ko]\n",
    "    N = kos_background[ko]\n",
    "    p_values[ko] = hypergeom.sf(k-1, M, N, n)\n",
    "    enrichments[ko] = (k/n) / (N/M)\n",
    "    enrichments[ko] = \"positivo\" if enrichments[ko] > 1 else \"negativo\"\n",
    "df=pd.DataFrame.from_dict(p_values,orient=\"index\",columns=[\"p_value\"])\n",
    "df[\"enrichment\"] = pd.Series(enrichments)\n",
    "for ko in kos:\n",
    "    df.at[ko, \"genes with motif\"] = \", \".join(ko_to_genes[ko])\n",
    "df.sort_values(\"p_value\", inplace=True, ascending=True)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ],
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2024-04-06T17:58:14.036695Z",
     "start_time": "2024-04-06T17:58:08.809559Z"
    }
   },
   "id": "initial_id",
   "execution_count": 7
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "                                                     p_value enrichment  \\\n03440 Homologous recombination                      0.109839   positivo   \n00280 Valine, leucine and isoleucine degradation    0.169068   positivo   \n03430 Mismatch repair                               0.255539   positivo   \n03030 DNA replication                               0.317363   positivo   \n00730 Thiamine metabolism                           0.366289   positivo   \n00640 Propanoate metabolism                         0.413816   positivo   \n00430 Taurine and hypotaurine metabolism            0.417648   positivo   \n00340 Histidine metabolism                          0.459517   positivo   \n00310 Lysine degradation                            0.491413   positivo   \n00380 Tryptophan metabolism                         0.612220   positivo   \n00660 C5-Branched dibasic acid metabolism           0.661448   negativo   \n00460 Cyanoamino acid metabolism                    0.704459   negativo   \n00552 Teichoic acid biosynthesis                    0.704459   negativo   \n00780 Biotin metabolism                             0.742033   negativo   \n02026 Biofilm formation - Escherichia coli          0.742033   negativo   \n00785 Lipoic acid metabolism                        0.742033   negativo   \n04122 Sulfur relay system                           0.803522   negativo   \n04112 Cell cycle - Caulobacter                      0.828557   negativo   \n00906 Carotenoid biosynthesis                       0.828557   negativo   \n00020 Citrate cycle                                 0.850419   negativo   \n00650 Butanoate metabolism                          0.869508   negativo   \n00290 Valine, leucine and isoleucine biosynthesis   0.869508   negativo   \n05111 Biofilm formation - Vibrio cholerae           0.886172   negativo   \n03070 Bacterial secretion system                    0.886172   negativo   \n01210 2-Oxocarboxylic acid metabolism               0.899017   negativo   \n00630 Glyoxylate and dicarboxylate metabolism       0.909526   negativo   \n00770 Pantothenate and CoA biosynthesis             0.913418   negativo   \n00480 Glutathione metabolism                        0.949971   negativo   \n00240 Pyrimidine metabolism                         0.956394   negativo   \n00260 Glycine, serine and threonine metabolism      0.966882   negativo   \n00720 Carbon fixation pathways in prokaryotes       0.966882   negativo   \n01232 Nucleotide metabolism                         0.966882   negativo   \n00970 Aminoacyl-tRNA biosynthesis                   0.971143   negativo   \n00620 Pyruvate metabolism                           0.980923   negativo   \n00010 Glycolysis / Gluconeogenesis                  0.980923   negativo   \n03010 Ribosome                                      0.994655   negativo   \n00541 O-Antigen nucleotide sugar biosynthesis       0.995233   negativo   \n00195 Photosynthesis                                0.998961   negativo   \n02020 Two-component system                          0.998961   negativo   \n01250 Biosynthesis of nucleotide sugars             0.999877   negativo   \n01200 Carbon metabolism                             0.999901   negativo   \n01230 Biosynthesis of amino acids                   0.999997   negativo   \n01240 Biosynthesis of cofactors                     1.000000   negativo   \n02010 ABC transporters                              1.000000   negativo   \n01120 Microbial metabolism in diverse environments  1.000000   negativo   \n01110 Biosynthesis of secondary metabolites         1.000000   negativo   \n01100 Metabolic pathways                            1.000000   negativo   \n\n                                                                                     genes with motif  \n03440 Homologous recombination                                                 None, recA, None, None  \n00280 Valine, leucine and isoleucine degradation                                           lpdA, None  \n03430 Mismatch repair                                                                None, xseA, None  \n03030 DNA replication                                                                      None, None  \n00730 Thiamine metabolism                                                                  None, thiD  \n00640 Propanoate metabolism                                                                lpdA, None  \n00430 Taurine and hypotaurine metabolism                                                          ggt  \n00340 Histidine metabolism                                                                 None, hisD  \n00310 Lysine degradation                                                                         lpdA  \n00380 Tryptophan metabolism                                                                      lpdA  \n00660 C5-Branched dibasic acid metabolism                                                        None  \n00460 Cyanoamino acid metabolism                                                                  ggt  \n00552 Teichoic acid biosynthesis                                                                 None  \n00780 Biotin metabolism                                                                          bioA  \n02026 Biofilm formation - Escherichia coli                                                       None  \n00785 Lipoic acid metabolism                                                                     lpdA  \n04122 Sulfur relay system                                                                        None  \n04112 Cell cycle - Caulobacter                                                                   None  \n00906 Carotenoid biosynthesis                                                                    crtH  \n00020 Citrate cycle                                                                              lpdA  \n00650 Butanoate metabolism                                                                       None  \n00290 Valine, leucine and isoleucine biosynthesis                                                None  \n05111 Biofilm formation - Vibrio cholerae                                                        None  \n03070 Bacterial secretion system                                                                 None  \n01210 2-Oxocarboxylic acid metabolism                                                      None, lpdA  \n00630 Glyoxylate and dicarboxylate metabolism                                              lpdA, None  \n00770 Pantothenate and CoA biosynthesis                                                          None  \n00480 Glutathione metabolism                                                                      ggt  \n00240 Pyrimidine metabolism                                                                       upp  \n00260 Glycine, serine and threonine metabolism                                                   lpdA  \n00720 Carbon fixation pathways in prokaryotes                                                    None  \n01232 Nucleotide metabolism                                                                       upp  \n00970 Aminoacyl-tRNA biosynthesis                                                                cysS  \n00620 Pyruvate metabolism                                                                        lpdA  \n00010 Glycolysis / Gluconeogenesis                                                               lpdA  \n03010 Ribosome                                                                             rplC, rpsT  \n00541 O-Antigen nucleotide sugar biosynthesis                                                    None  \n00195 Photosynthesis                                                                       None, psbD  \n02020 Two-component system                                                                 None, None  \n01250 Biosynthesis of nucleotide sugars                                                          None  \n01200 Carbon metabolism                                                                    lpdA, None  \n01230 Biosynthesis of amino acids                                                          None, hisD  \n01240 Biosynthesis of cofactors                                                lpdA, None, bioA, thiD  \n02010 ABC transporters                                                                           None  \n01120 Microbial metabolism in diverse environments                                         lpdA, None  \n01110 Biosynthesis of secondary metabolites                              None, lpdA, crtH, None, hisD  \n01100 Metabolic pathways                            None, None, None, lpdA, crtH, upp, None, None,...  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>p_value</th>\n      <th>enrichment</th>\n      <th>genes with motif</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>03440 Homologous recombination</th>\n      <td>0.109839</td>\n      <td>positivo</td>\n      <td>None, recA, None, None</td>\n    </tr>\n    <tr>\n      <th>00280 Valine, leucine and isoleucine degradation</th>\n      <td>0.169068</td>\n      <td>positivo</td>\n      <td>lpdA, None</td>\n    </tr>\n    <tr>\n      <th>03430 Mismatch repair</th>\n      <td>0.255539</td>\n      <td>positivo</td>\n      <td>None, xseA, None</td>\n    </tr>\n    <tr>\n      <th>03030 DNA replication</th>\n      <td>0.317363</td>\n      <td>positivo</td>\n      <td>None, None</td>\n    </tr>\n    <tr>\n      <th>00730 Thiamine metabolism</th>\n      <td>0.366289</td>\n      <td>positivo</td>\n      <td>None, thiD</td>\n    </tr>\n    <tr>\n      <th>00640 Propanoate metabolism</th>\n      <td>0.413816</td>\n      <td>positivo</td>\n      <td>lpdA, None</td>\n    </tr>\n    <tr>\n      <th>00430 Taurine and hypotaurine metabolism</th>\n      <td>0.417648</td>\n      <td>positivo</td>\n      <td>ggt</td>\n    </tr>\n    <tr>\n      <th>00340 Histidine metabolism</th>\n      <td>0.459517</td>\n      <td>positivo</td>\n      <td>None, hisD</td>\n    </tr>\n    <tr>\n      <th>00310 Lysine degradation</th>\n      <td>0.491413</td>\n      <td>positivo</td>\n      <td>lpdA</td>\n    </tr>\n    <tr>\n      <th>00380 Tryptophan metabolism</th>\n      <td>0.612220</td>\n      <td>positivo</td>\n      <td>lpdA</td>\n    </tr>\n    <tr>\n      <th>00660 C5-Branched dibasic acid metabolism</th>\n      <td>0.661448</td>\n      <td>negativo</td>\n      <td>None</td>\n    </tr>\n    <tr>\n      <th>00460 Cyanoamino acid metabolism</th>\n      <td>0.704459</td>\n      <td>negativo</td>\n      <td>ggt</td>\n    </tr>\n    <tr>\n      <th>00552 Teichoic acid biosynthesis</th>\n      <td>0.704459</td>\n      <td>negativo</td>\n      <td>None</td>\n    </tr>\n    <tr>\n      <th>00780 Biotin metabolism</th>\n      <td>0.742033</td>\n      <td>negativo</td>\n      <td>bioA</td>\n    </tr>\n    <tr>\n      <th>02026 Biofilm formation - Escherichia coli</th>\n      <td>0.742033</td>\n      <td>negativo</td>\n      <td>None</td>\n    </tr>\n    <tr>\n      <th>00785 Lipoic acid metabolism</th>\n      <td>0.742033</td>\n      <td>negativo</td>\n      <td>lpdA</td>\n    </tr>\n    <tr>\n      <th>04122 Sulfur relay system</th>\n      <td>0.803522</td>\n      <td>negativo</td>\n      <td>None</td>\n    </tr>\n    <tr>\n      <th>04112 Cell cycle - Caulobacter</th>\n      <td>0.828557</td>\n      <td>negativo</td>\n      <td>None</td>\n    </tr>\n    <tr>\n      <th>00906 Carotenoid biosynthesis</th>\n      <td>0.828557</td>\n      <td>negativo</td>\n      <td>crtH</td>\n    </tr>\n    <tr>\n      <th>00020 Citrate cycle</th>\n      <td>0.850419</td>\n      <td>negativo</td>\n      <td>lpdA</td>\n    </tr>\n    <tr>\n      <th>00650 Butanoate metabolism</th>\n      <td>0.869508</td>\n      <td>negativo</td>\n      <td>None</td>\n    </tr>\n    <tr>\n      <th>00290 Valine, leucine and isoleucine biosynthesis</th>\n      <td>0.869508</td>\n      <td>negativo</td>\n      <td>None</td>\n    </tr>\n    <tr>\n      <th>05111 Biofilm formation - Vibrio cholerae</th>\n      <td>0.886172</td>\n      <td>negativo</td>\n      <td>None</td>\n    </tr>\n    <tr>\n      <th>03070 Bacterial secretion system</th>\n      <td>0.886172</td>\n      <td>negativo</td>\n      <td>None</td>\n    </tr>\n    <tr>\n      <th>01210 2-Oxocarboxylic acid metabolism</th>\n      <td>0.899017</td>\n      <td>negativo</td>\n      <td>None, lpdA</td>\n    </tr>\n    <tr>\n      <th>00630 Glyoxylate and dicarboxylate metabolism</th>\n      <td>0.909526</td>\n      <td>negativo</td>\n      <td>lpdA, None</td>\n    </tr>\n    <tr>\n      <th>00770 Pantothenate and CoA biosynthesis</th>\n      <td>0.913418</td>\n      <td>negativo</td>\n      <td>None</td>\n    </tr>\n    <tr>\n      <th>00480 Glutathione metabolism</th>\n      <td>0.949971</td>\n      <td>negativo</td>\n      <td>ggt</td>\n    </tr>\n    <tr>\n      <th>00240 Pyrimidine metabolism</th>\n      <td>0.956394</td>\n      <td>negativo</td>\n      <td>upp</td>\n    </tr>\n    <tr>\n      <th>00260 Glycine, serine and threonine metabolism</th>\n      <td>0.966882</td>\n      <td>negativo</td>\n      <td>lpdA</td>\n    </tr>\n    <tr>\n      <th>00720 Carbon fixation pathways in prokaryotes</th>\n      <td>0.966882</td>\n      <td>negativo</td>\n      <td>None</td>\n    </tr>\n    <tr>\n      <th>01232 Nucleotide metabolism</th>\n      <td>0.966882</td>\n      <td>negativo</td>\n      <td>upp</td>\n    </tr>\n    <tr>\n      <th>00970 Aminoacyl-tRNA biosynthesis</th>\n      <td>0.971143</td>\n      <td>negativo</td>\n      <td>cysS</td>\n    </tr>\n    <tr>\n      <th>00620 Pyruvate metabolism</th>\n      <td>0.980923</td>\n      <td>negativo</td>\n      <td>lpdA</td>\n    </tr>\n    <tr>\n      <th>00010 Glycolysis / Gluconeogenesis</th>\n      <td>0.980923</td>\n      <td>negativo</td>\n      <td>lpdA</td>\n    </tr>\n    <tr>\n      <th>03010 Ribosome</th>\n      <td>0.994655</td>\n      <td>negativo</td>\n      <td>rplC, rpsT</td>\n    </tr>\n    <tr>\n      <th>00541 O-Antigen nucleotide sugar biosynthesis</th>\n      <td>0.995233</td>\n      <td>negativo</td>\n      <td>None</td>\n    </tr>\n    <tr>\n      <th>00195 Photosynthesis</th>\n      <td>0.998961</td>\n      <td>negativo</td>\n      <td>None, psbD</td>\n    </tr>\n    <tr>\n      <th>02020 Two-component system</th>\n      <td>0.998961</td>\n      <td>negativo</td>\n      <td>None, None</td>\n    </tr>\n    <tr>\n      <th>01250 Biosynthesis of nucleotide sugars</th>\n      <td>0.999877</td>\n      <td>negativo</td>\n      <td>None</td>\n    </tr>\n    <tr>\n      <th>01200 Carbon metabolism</th>\n      <td>0.999901</td>\n      <td>negativo</td>\n      <td>lpdA, None</td>\n    </tr>\n    <tr>\n      <th>01230 Biosynthesis of amino acids</th>\n      <td>0.999997</td>\n      <td>negativo</td>\n      <td>None, hisD</td>\n    </tr>\n    <tr>\n      <th>01240 Biosynthesis of cofactors</th>\n      <td>1.000000</td>\n      <td>negativo</td>\n      <td>lpdA, None, bioA, thiD</td>\n    </tr>\n    <tr>\n      <th>02010 ABC transporters</th>\n      <td>1.000000</td>\n      <td>negativo</td>\n      <td>None</td>\n    </tr>\n    <tr>\n      <th>01120 Microbial metabolism in diverse environments</th>\n      <td>1.000000</td>\n      <td>negativo</td>\n      <td>lpdA, None</td>\n    </tr>\n    <tr>\n      <th>01110 Biosynthesis of secondary metabolites</th>\n      <td>1.000000</td>\n      <td>negativo</td>\n      <td>None, lpdA, crtH, None, hisD</td>\n    </tr>\n    <tr>\n      <th>01100 Metabolic pathways</th>\n      <td>1.000000</td>\n      <td>negativo</td>\n      <td>None, None, None, lpdA, crtH, upp, None, None,...</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-04T16:13:30.218765Z",
     "start_time": "2024-04-04T16:13:30.198022Z"
    }
   },
   "id": "d1a4e112dc35201",
   "execution_count": 23
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "171 1183\n"
     ]
    }
   ],
   "source": [
    "print(i,j)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-04T16:15:47.901149Z",
     "start_time": "2024-04-04T16:15:47.898840Z"
    }
   },
   "id": "285d3771f78835cd",
   "execution_count": 24
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "Counter({'01100 Metabolic pathways': 14,\n         '01110 Biosynthesis of secondary metabolites': 5,\n         '03440 Homologous recombination': 4,\n         '01240 Biosynthesis of cofactors': 4,\n         '03430 Mismatch repair': 3,\n         '03030 DNA replication': 2,\n         '00195 Photosynthesis': 2,\n         '01210 2-Oxocarboxylic acid metabolism': 2,\n         '01230 Biosynthesis of amino acids': 2,\n         '02020 Two-component system': 2,\n         '01120 Microbial metabolism in diverse environments': 2,\n         '01200 Carbon metabolism': 2,\n         '00630 Glyoxylate and dicarboxylate metabolism': 2,\n         '00640 Propanoate metabolism': 2,\n         '00280 Valine, leucine and isoleucine degradation': 2,\n         '00730 Thiamine metabolism': 2,\n         '00340 Histidine metabolism': 2,\n         '03010 Ribosome': 2,\n         '05111 Biofilm formation - Vibrio cholerae': 1,\n         '02026 Biofilm formation - Escherichia coli': 1,\n         '01250 Biosynthesis of nucleotide sugars': 1,\n         '00541 O-Antigen nucleotide sugar biosynthesis': 1,\n         '00552 Teichoic acid biosynthesis': 1,\n         '00650 Butanoate metabolism': 1,\n         '00660 C5-Branched dibasic acid metabolism': 1,\n         '00290 Valine, leucine and isoleucine biosynthesis': 1,\n         '00770 Pantothenate and CoA biosynthesis': 1,\n         '00010 Glycolysis / Gluconeogenesis': 1,\n         '00020 Citrate cycle': 1,\n         '00620 Pyruvate metabolism': 1,\n         '00260 Glycine, serine and threonine metabolism': 1,\n         '00310 Lysine degradation': 1,\n         '00380 Tryptophan metabolism': 1,\n         '00785 Lipoic acid metabolism': 1,\n         '00906 Carotenoid biosynthesis': 1,\n         '01232 Nucleotide metabolism': 1,\n         '00240 Pyrimidine metabolism': 1,\n         '00720 Carbon fixation pathways in prokaryotes': 1,\n         '04122 Sulfur relay system': 1,\n         '00430 Taurine and hypotaurine metabolism': 1,\n         '00460 Cyanoamino acid metabolism': 1,\n         '00480 Glutathione metabolism': 1,\n         '04112 Cell cycle - Caulobacter': 1,\n         '00970 Aminoacyl-tRNA biosynthesis': 1,\n         '00780 Biotin metabolism': 1,\n         '02010 ABC transporters': 1,\n         '03070 Bacterial secretion system': 1})"
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kos"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-06T17:58:22.193922Z",
     "start_time": "2024-04-06T17:58:22.188599Z"
    }
   },
   "id": "35ab9bcdc9a2b12f",
   "execution_count": 8
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "Counter({'01100 Metabolic pathways': 760,\n         '01110 Biosynthesis of secondary metabolites': 346,\n         '01240 Biosynthesis of cofactors': 158,\n         '01120 Microbial metabolism in diverse environments': 157,\n         '02010 ABC transporters': 117,\n         '01230 Biosynthesis of amino acids': 111,\n         '01200 Carbon metabolism': 85,\n         '02020 Two-component system': 67,\n         '00195 Photosynthesis': 67,\n         '01250 Biosynthesis of nucleotide sugars': 65,\n         '03010 Ribosome': 54,\n         '00860 Porphyrin metabolism': 51,\n         '00520 Amino sugar and nucleotide sugar metabolism': 43,\n         '00190 Oxidative phosphorylation': 42,\n         '00230 Purine metabolism': 40,\n         '00541 O-Antigen nucleotide sugar biosynthesis': 39,\n         '02024 Quorum sensing': 35,\n         '00500 Starch and sucrose metabolism': 32,\n         '00270 Cysteine and methionine metabolism': 30,\n         '00630 Glyoxylate and dicarboxylate metabolism': 30,\n         '00400 Phenylalanine, tyrosine and tryptophan biosynthesis': 29,\n         '00010 Glycolysis / Gluconeogenesis': 29,\n         '01210 2-Oxocarboxylic acid metabolism': 29,\n         '00620 Pyruvate metabolism': 29,\n         '00790 Folate biosynthesis': 28,\n         '00970 Aminoacyl-tRNA biosynthesis': 26,\n         '00260 Glycine, serine and threonine metabolism': 25,\n         '00030 Pentose phosphate pathway': 25,\n         '00720 Carbon fixation pathways in prokaryotes': 25,\n         '00680 Methane metabolism': 25,\n         '01232 Nucleotide metabolism': 25,\n         '00240 Pyrimidine metabolism': 23,\n         '00910 Nitrogen metabolism': 23,\n         '00250 Alanine, aspartate and glutamate metabolism': 23,\n         '00480 Glutathione metabolism': 22,\n         '00920 Sulfur metabolism': 22,\n         '00521 Streptomycin biosynthesis': 22,\n         '00130 Ubiquinone and other terpenoid-quinone biosynthesis': 22,\n         '00550 Peptidoglycan biosynthesis': 21,\n         '00051 Fructose and mannose metabolism': 20,\n         '00760 Nicotinate and nicotinamide metabolism': 19,\n         '00523 Polyketide sugar unit biosynthesis': 19,\n         '01212 Fatty acid metabolism': 19,\n         '00983 Drug metabolism - other enzymes': 18,\n         '00710 Carbon fixation in photosynthetic organisms': 18,\n         '03018 RNA degradation': 18,\n         '00770 Pantothenate and CoA biosynthesis': 18,\n         '00220 Arginine biosynthesis': 18,\n         '00330 Arginine and proline metabolism': 17,\n         '03060 Protein export': 17,\n         '05111 Biofilm formation - Vibrio cholerae': 16,\n         '03070 Bacterial secretion system': 16,\n         '00290 Valine, leucine and isoleucine biosynthesis': 15,\n         '00061 Fatty acid biosynthesis': 15,\n         '00650 Butanoate metabolism': 15,\n         '00196 Photosynthesis - antenna proteins': 15,\n         '03440 Homologous recombination': 15,\n         '00020 Citrate cycle': 14,\n         '00900 Terpenoid backbone biosynthesis': 14,\n         '03430 Mismatch repair': 14,\n         '00906 Carotenoid biosynthesis': 13,\n         '00561 Glycerolipid metabolism': 13,\n         '04112 Cell cycle - Caulobacter': 13,\n         '00300 Lysine biosynthesis': 12,\n         '04122 Sulfur relay system': 12,\n         '00350 Tyrosine metabolism': 12,\n         '00670 One carbon pool by folate': 12,\n         '00540 Lipopolysaccharide biosynthesis': 12,\n         '00340 Histidine metabolism': 12,\n         '00360 Phenylalanine metabolism': 12,\n         '00980 Metabolism of xenobiotics by cytochrome P450': 11,\n         '00982 Drug metabolism - cytochrome P450': 11,\n         '00640 Propanoate metabolism': 11,\n         '03410 Base excision repair': 11,\n         '00785 Lipoic acid metabolism': 10,\n         '00730 Thiamine metabolism': 10,\n         '00780 Biotin metabolism': 10,\n         '02026 Biofilm formation - Escherichia coli': 10,\n         '00052 Galactose metabolism': 10,\n         '04146 Peroxisome': 9,\n         '00552 Teichoic acid biosynthesis': 9,\n         '00470 D-Amino acid metabolism': 9,\n         '00071 Fatty acid degradation': 9,\n         '00525 Acarbose and validamycin biosynthesis': 9,\n         '00460 Cyanoamino acid metabolism': 9,\n         '03030 DNA replication': 9,\n         '00740 Riboflavin metabolism': 8,\n         '03420 Nucleotide excision repair': 8,\n         '00040 Pentose and glucuronate interconversions': 8,\n         '00660 C5-Branched dibasic acid metabolism': 8,\n         '02030 Bacterial chemotaxis': 8,\n         '00627 Aminobenzoate degradation': 7,\n         '00380 Tryptophan metabolism': 7,\n         '00543 Exopolysaccharide biosynthesis': 7,\n         '02025 Biofilm formation - Pseudomonas aeruginosa': 7,\n         '00362 Benzoate degradation': 7,\n         '00564 Glycerophospholipid metabolism': 7,\n         '00450 Selenocompound metabolism': 6,\n         '04217 Necroptosis': 6,\n         '00280 Valine, leucine and isoleucine degradation': 6,\n         '00261 Monobactam biosynthesis': 6,\n         '00750 Vitamin B6 metabolism': 6,\n         '04066 HIF-1 signaling pathway': 5,\n         '00053 Ascorbate and aldarate metabolism': 5,\n         '00310 Lysine degradation': 5,\n         '00999 Biosynthesis of various plant secondary metabolites; Including: Crocin biosynthesis, Ginsenoside biosynthesis, Saponin adjuvant biosynthesis, Cannabidiol biosynthesis, Mugineic acid biosynthesis, Pentagalloylglucose biosynthesis, Benzoxazinoid biosynthesis, Gramine biosynthesis, Coumarin biosynthesis, Furanocoumarin biosynthesis, Hordatine biosynthesis, Podophyllotoxin biosynthesis\\xa0(3)K00789WP_250123532.1, WP_250124149.1, WP_250125002.1K05349WP_250123179.1K05350WP_250125274.1  Xenobiotics biodegradation and metabolism  00362 Benzoate degradation\\xa0(6)K00074WP_250125631.1, WP_250126457.1K00626WP_250121756.1K00632WP_250121366.1K01607WP_250122318.1K04110WP_250126464.1K07516WP_250121367.100627 Aminobenzoate degradation\\xa0(5)K01426WP_250123433.1, WP_250126116.1, WP_250126722.1K01512WP_250121171.1K01576WP_250122564.1K03186WP_250124261.1K04110WP_250126464.100364 Fluorobenzoate degradation\\xa0(1)K01061WP_250124637.1, WP_250124906.1, WP_250126377.1, WP_250126683.1, WP_250126736.100625 Chloroalkane and chloroalkene degradation\\xa0(2)K00128WP_250124621.1K13953WP_250121168.100361 Chlorocyclohexane and chlorobenzene degradation\\xa0(1)K01061WP_250124637.1, WP_250124906.1, WP_250126377.1, WP_250126683.1, WP_250126736.100623 Toluene degradation\\xa0(1)K01061WP_250124637.1, WP_250124906.1, WP_250126377.1, WP_250126683.1, WP_250126736.100642 Ethylbenzene degradation\\xa0(1)K00632WP_250121366.100643 Styrene degradation\\xa0(3)K00451WP_250121340.1K01426WP_250123433.1, WP_250126116.1, WP_250126722.1K01555WP_250122169.100791 Atrazine degradation\\xa0(3)K01428WP_250124618.1K01429WP_250124617.1K01430WP_250124616.100930 Caprolactam degradation\\xa0(1)K18199WP_250123455.1, WP_250124848.1, WP_250125744.100626 Naphthalene degradation\\xa0(1)K13953WP_250121168.100980 Metabolism of xenobiotics by cytochrome P450\\xa0(2)K00799WP_250121179.1, WP_250121400.1, WP_250121404.1, WP_250122618.1, WP_250123314.1, WP_250123457.1, WP_250123587.1, WP_250123658.1, WP_250126707.1, WP_256478909.1K13953WP_250121168.100982 Drug metabolism - cytochrome P450\\xa0(2)K00799WP_250121179.1, WP_250121400.1, WP_250121404.1, WP_250122618.1, WP_250123314.1, WP_250123457.1, WP_250123587.1, WP_250123658.1, WP_250126707.1, WP_256478909.1K13953WP_250121168.100983 Drug metabolism - other enzymes\\xa0(8)K00088WP_250123285.1K00757WP_250125855.1K00799WP_250121179.1, WP_250121400.1, WP_250121404.1, WP_250122618.1, WP_250123314.1, WP_250123457.1, WP_250123587.1, WP_250123658.1, WP_250126707.1, WP_256478909.1K00940WP_250124804.1K01519WP_250122237.1K01951WP_250125777.1, WP_275983373.1K03782WP_250122924.1K13421WP_250123530.1 Genetic Information Processing Transcription  03020 RNA polymerase\\xa0(3)K03040WP_250125971.1K03043WP_250123319.1K03060WP_250122201.103022 Basal transcription factors\\xa0(1)K10843WP_250126525.1  Translation  03010 Ribosome\\xa0(53)K02863WP_250125426.1K02864WP_250125425.1K02867WP_250125427.1K02871WP_250125968.1K02874WP_250125983.1K02876WP_250125977.1K02878WP_250125986.1K02879WP_250125970.1K02881WP_250125979.1K02884WP_250125430.1K02886WP_250125990.1K02887WP_250122606.1K02888WP_250125552.1K02890WP_250125988.1K02892WP_250125991.1K02895WP_250126203.1K02897WP_250124527.1K02899WP_250125553.1K02902WP_250122722.1K02904WP_250125985.1K02906WP_250125993.1K02909WP_250125966.1K02911WP_250124624.1K02913WP_250121863.1K02914WP_250123140.1K02916WP_250122605.1K02919WP_250125974.1K02926WP_250125992.1K02931WP_250125982.1K02933WP_250125980.1K02935WP_250125424.1K02939WP_250121540.1K02945WP_250123547.1, WP_250124638.1K02946WP_250121290.1K02948WP_250125972.1K02950WP_250121286.1K02952WP_250125973.1K02954WP_250121552.1K02956WP_250125336.1K02959WP_250124981.1K02961WP_250125984.1K02963WP_250121862.1K02965WP_250125989.1K02967WP_250124818.1K02968WP_250123321.1K02970WP_250122015.1, WP_250122274.1, WP_250122333.1K02982WP_250125987.1K02986WP_250124119.1K02988WP_250125978.1K02990WP_250122951.1K02992WP_250121287.1K02994WP_250125981.1K02996WP_250125967.100970 Aminoacyl-tRNA biosynthesis\\xa0(25)K00604WP_250125215.1K01866WP_250124730.1K01867WP_250126166.1K01868WP_250125326.1K01869WP_250125368.1K01870WP_250123615.1K01872WP_250123557.1K01873WP_250124339.1K01874WP_250123267.1K01875WP_250121555.1K01876WP_250122320.1K01878WP_250121451.1K01879WP_250121991.1K01881WP_250124443.1K01883WP_250121278.1, WP_250126501.1K01885WP_250123393.1K01887WP_250124329.1K01889WP_250123149.1K01890WP_250121868.1K01892WP_250122185.1K01893WP_250124789.1K02433WP_250121584.1, WP_250122150.1K02434WP_250124516.1K02435WP_250122464.1K04567WP_250121498.103008 Ribosome biogenesis in eukaryotes\\xa0(1)K03685WP_250122345.1  Folding, sorting and degradation  03060 Protein export\\xa0(14)K01551WP_250125151.1K03070WP_250123145.1K03072WP_250124360.1K03073WP_250125429.1K03074WP_250124361.1K03075WP_250125958.1K03076WP_250125976.1K03100WP_250121549.1, WP_250122056.1, WP_250124735.1K03101WP_250123095.1K03106WP_250124982.1K03110WP_250125439.1K03116WP_250122331.1, WP_250123883.1K03118WP_250121784.1K03217WP_256478733.104141 Protein processing in endoplasmic reticulum\\xa0(2)K04079WP_250122721.1K13993WP_250125344.1, WP_250126071.1, WP_250126072.104122 Sulfur relay system\\xa0(10)K00566WP_250121598.1K01011WP_250122755.1, WP_250123031.1K03154WP_250121543.1K03635WP_250123857.1K03636WP_250122363.1, WP_250122366.1K03637WP_250122910.1K03638WP_250121271.1K03639WP_250124120.1K04487WP_250125134.1, WP_250125883.1K21147WP_250125482.103018 RNA degradation\\xa0(13)K00937WP_250121710.1K00962WP_250122897.1K01689WP_250125716.1K03654WP_250125213.1, WP_250125324.1K03666WP_250123766.1K04043WP_250122003.1, WP_250122389.1, WP_250122543.1, WP_250126349.1K04077WP_250122604.1, WP_250124781.1K05592WP_250122989.1K08300WP_250121295.1K11927WP_250121888.1K12573WP_250124262.1K12574WP_250124627.1K22468WP_250122842.1  Replication and repair  03030 DNA replication\\xa0(11)K01972WP_250121679.1K02314WP_250121539.1K02316WP_250122123.1K02337WP_250122145.1K02338WP_250125153.1K02340WP_250121983.1K02341WP_250122066.1K02343WP_250123933.1K03111WP_250122517.1K03469WP_250125557.1K03470WP_250121294.103410 Base excision repair\\xa0(10)K01142WP_250126015.1K01247WP_250124583.1K01972WP_250121679.1K03575WP_250124604.1K03649WP_250125234.1K03652WP_250125868.1K07462WP_250121314.1, WP_250125167.1K10563WP_250126148.1K10773WP_250121553.1K21929WP_250122964.103420 Nucleotide excision repair\\xa0(7)K01972WP_250121679.1K03657WP_250124171.1, WP_250125591.1K03701WP_250122461.1K03702WP_250123858.1K03703WP_250126083.1K03723WP_250122433.1K10843WP_250126525.103430 Mismatch repair\\xa0(15)K01972WP_250121679.1K02337WP_250122145.1K02338WP_250125153.1K02340WP_250121983.1K02341WP_250122066.1K02343WP_250123933.1K03111WP_250122517.1K03555WP_250121965.1K03572WP_250123089.1K03601WP_250121492.1K03602WP_250121489.1K03657WP_250124171.1, WP_250125591.1K06223WP_250123176.1K07456WP_250123284.1K07462WP_250121314.1, WP_250125167.103440 Homologous recombination\\xa0(17)K01159WP_250121455.1K02337WP_250122145.1K02338WP_250125153.1K02340WP_250121983.1K02341WP_250122066.1K02343WP_250123933.1K03111WP_250122517.1K03550WP_250123053.1K03551WP_250122501.1K03553WP_250121493.1K03581WP_250121656.1, WP_250126637.1K03584WP_250125952.1K03629WP_250121880.1K03655WP_250124815.1K04066WP_250122904.1K06187WP_250126200.1K07462WP_250121314.1, WP_250125167.1 Environmental Information Processing Membrane transport  02010 ABC transporters\\xa0(89)K01995WP_250124133.1K01996WP_250121449.1K01997WP_250122242.1K01998WP_250124132.1K01999WP_250124166.1K02006WP_250125255.1K02007WP_250125253.1K02008WP_250125254.1K02010WP_250121157.1K02011WP_250121158.1, WP_250122165.1K02012WP_250122542.1K02018WP_250122859.1K02020WP_250122860.1K02036WP_250123093.1, WP_250123396.1K02037WP_250123091.1, WP_250123398.1K02038WP_250123092.1, WP_250123397.1K02040WP_250122421.1, WP_250123074.1, WP_250123090.1, WP_250123399.1, WP_250124113.1, WP_250124114.1K02045WP_250121827.1, WP_250123699.1, WP_250126491.1K02046WP_250124787.1K02047WP_250124786.1K02065WP_250124734.1K02066WP_250126101.1K02067WP_250124733.1K03523WP_250123094.1K05845WP_250125189.1K05846WP_250122140.1K05847WP_250122139.1K06160WP_250121614.1K06861WP_250122291.1K09690WP_250122998.1, WP_250125032.1, WP_250125060.1, WP_250125082.1, WP_250125300.1K09691WP_250123000.1, WP_250124651.1, WP_250125031.1, WP_250125059.1, WP_250125080.1, WP_250125301.1K09815WP_250125474.1K09816WP_250125475.1K09817WP_250126190.1K09969WP_250121977.1K09970WP_250121976.1K09971WP_250121975.1K09972WP_250126225.1K10112WP_250124502.1, WP_250124694.1, WP_275983366.1K10236WP_250124697.1K10237WP_250124696.1K10238WP_250124695.1K10440WP_250124562.1K10820WP_250124563.1K10829WP_250124555.1K11004WP_250122283.1, WP_250123361.1K11069WP_250123648.1K11070WP_250124276.1K11071WP_250123649.1K11072WP_250123647.1K11085WP_250121860.1, WP_250122637.1, WP_250125085.1, WP_250126168.1K11720WP_250122292.1K11950WP_250121845.1K11951WP_250121844.1K11952WP_250121843.1, WP_250122847.1K11953WP_250121842.1K11954WP_250122546.1K11955WP_250121988.1K11956WP_250125405.1K11957WP_250121987.1K11958WP_250125484.1K11959WP_250125386.1K11960WP_250125387.1K11961WP_250125388.1K11962WP_250125389.1K11963WP_250125391.1K14698WP_250126440.1K14699WP_250126446.1K15576WP_250122849.1, WP_250123888.1K15577WP_250122848.1, WP_250123887.1K15578WP_250123889.1K15579WP_250122846.1K16785WP_250125579.1K16786WP_250126053.1K16787WP_250121331.1K16915WP_250125252.1K17062WP_250124274.1K17063WP_250124275.1K17244WP_250122258.1K17245WP_250123326.1K17246WP_250126134.1K18233WP_250123152.1K18889WP_250126121.1K18890WP_250125106.1K23163WP_250124788.1K23227WP_250124556.1K23228WP_250124557.1K23508WP_250124564.1K23509WP_250124561.1': 5,\n         '00364 Fluorobenzoate degradation': 5,\n         '00361 Chlorocyclohexane and chlorobenzene degradation': 5,\n         '00623 Toluene degradation': 5,\n         '00643 Styrene degradation': 5,\n         '00405 Phenazine biosynthesis': 4,\n         '01053 Biosynthesis of siderophore group nonribosomal peptides': 4,\n         '00401 Novobiocin biosynthesis': 4,\n         '00410 beta-Alanine metabolism': 4,\n         '01055 Biosynthesis of vancomycin group antibiotics': 4,\n         '04214 Apoptosis - fly': 4,\n         '00430 Taurine and hypotaurine metabolism': 4,\n         '04141 Protein processing in endoplasmic reticulum': 4,\n         '04070 Phosphatidylinositol signaling system': 4,\n         '01220 Degradation of aromatic compounds': 3,\n         '00960 Tropane, piperidine and pyridine alkaloid biosynthesis': 3,\n         '00791 Atrazine degradation': 3,\n         '00966 Glucosinolate biosynthesis': 3,\n         '00930 Caprolactam degradation': 3,\n         '04080 Neuroactive ligand-receptor interaction': 3,\n         '00562 Inositol phosphate metabolism': 3,\n         '00510 N-Glycan biosynthesis': 3,\n         '04216 Ferroptosis': 3,\n         '03020 RNA polymerase': 3,\n         '00333 Prodigiosin biosynthesis': 3,\n         '00332 Carbapenem biosynthesis': 3,\n         '00940 Phenylpropanoid biosynthesis': 2,\n         '00625 Chloroalkane and chloroalkene degradation': 2,\n         '04016 MAPK signaling pathway - plant': 2,\n         '00907 Pinene, camphor and geraniol degradation': 2,\n         '00946 Degradation of flavonoids': 2,\n         '00908 Zeatin biosynthesis': 2,\n         '00524 Neomycin, kanamycin and gentamicin biosynthesis': 2,\n         '00511 Other glycan degradation': 2,\n         '00542 O-Antigen repeat unit biosynthesis': 2,\n         '00572 Arabinogalactan biosynthesis - Mycobacterium': 2,\n         '04013 MAPK signaling pathway - fly': 2,\n         '04068 FoxO signaling pathway': 2,\n         '00965 Betalain biosynthesis': 2,\n         '00950 Isoquinoline alkaloid biosynthesis': 2,\n         '01051 Biosynthesis of ansamycins': 2,\n         '00998 Biosynthesis of various antibiotics; Including: Kanosamine biosynthesis, Aurachin biosynthesis, Bacilysin biosynthesis, Puromycin biosynthesis, Dapdiamides biosynthesis, Fosfomycin biosynthesis, Cremeomycin biosynthesis, Fumagillin biosynthesis, Pentalenolactone biosynthesis, Terpentecin biosynthesis, Roseoflavin biosynthesis, Cycloserine biosynthesis\\xa0(1)K02078WP_250122401.1, WP_250126466.100999 Biosynthesis of various plant secondary metabolites; Including: Crocin biosynthesis, Ginsenoside biosynthesis, Saponin adjuvant biosynthesis, Cannabidiol biosynthesis, Mugineic acid biosynthesis, Pentagalloylglucose biosynthesis, Benzoxazinoid biosynthesis, Gramine biosynthesis, Coumarin biosynthesis, Furanocoumarin biosynthesis, Hordatine biosynthesis, Podophyllotoxin biosynthesis\\xa0(3)K00789WP_250123532.1, WP_250124149.1, WP_250125002.1K05349WP_250123179.1K05350WP_250125274.1  Xenobiotics biodegradation and metabolism  00362 Benzoate degradation\\xa0(6)K00074WP_250125631.1, WP_250126457.1K00626WP_250121756.1K00632WP_250121366.1K01607WP_250122318.1K04110WP_250126464.1K07516WP_250121367.100627 Aminobenzoate degradation\\xa0(5)K01426WP_250123433.1, WP_250126116.1, WP_250126722.1K01512WP_250121171.1K01576WP_250122564.1K03186WP_250124261.1K04110WP_250126464.100364 Fluorobenzoate degradation\\xa0(1)K01061WP_250124637.1, WP_250124906.1, WP_250126377.1, WP_250126683.1, WP_250126736.100625 Chloroalkane and chloroalkene degradation\\xa0(2)K00128WP_250124621.1K13953WP_250121168.100361 Chlorocyclohexane and chlorobenzene degradation\\xa0(1)K01061WP_250124637.1, WP_250124906.1, WP_250126377.1, WP_250126683.1, WP_250126736.100623 Toluene degradation\\xa0(1)K01061WP_250124637.1, WP_250124906.1, WP_250126377.1, WP_250126683.1, WP_250126736.100642 Ethylbenzene degradation\\xa0(1)K00632WP_250121366.100643 Styrene degradation\\xa0(3)K00451WP_250121340.1K01426WP_250123433.1, WP_250126116.1, WP_250126722.1K01555WP_250122169.100791 Atrazine degradation\\xa0(3)K01428WP_250124618.1K01429WP_250124617.1K01430WP_250124616.100930 Caprolactam degradation\\xa0(1)K18199WP_250123455.1, WP_250124848.1, WP_250125744.100626 Naphthalene degradation\\xa0(1)K13953WP_250121168.100980 Metabolism of xenobiotics by cytochrome P450\\xa0(2)K00799WP_250121179.1, WP_250121400.1, WP_250121404.1, WP_250122618.1, WP_250123314.1, WP_250123457.1, WP_250123587.1, WP_250123658.1, WP_250126707.1, WP_256478909.1K13953WP_250121168.100982 Drug metabolism - cytochrome P450\\xa0(2)K00799WP_250121179.1, WP_250121400.1, WP_250121404.1, WP_250122618.1, WP_250123314.1, WP_250123457.1, WP_250123587.1, WP_250123658.1, WP_250126707.1, WP_256478909.1K13953WP_250121168.100983 Drug metabolism - other enzymes\\xa0(8)K00088WP_250123285.1K00757WP_250125855.1K00799WP_250121179.1, WP_250121400.1, WP_250121404.1, WP_250122618.1, WP_250123314.1, WP_250123457.1, WP_250123587.1, WP_250123658.1, WP_250126707.1, WP_256478909.1K00940WP_250124804.1K01519WP_250122237.1K01951WP_250125777.1, WP_275983373.1K03782WP_250122924.1K13421WP_250123530.1 Genetic Information Processing Transcription  03020 RNA polymerase\\xa0(3)K03040WP_250125971.1K03043WP_250123319.1K03060WP_250122201.103022 Basal transcription factors\\xa0(1)K10843WP_250126525.1  Translation  03010 Ribosome\\xa0(53)K02863WP_250125426.1K02864WP_250125425.1K02867WP_250125427.1K02871WP_250125968.1K02874WP_250125983.1K02876WP_250125977.1K02878WP_250125986.1K02879WP_250125970.1K02881WP_250125979.1K02884WP_250125430.1K02886WP_250125990.1K02887WP_250122606.1K02888WP_250125552.1K02890WP_250125988.1K02892WP_250125991.1K02895WP_250126203.1K02897WP_250124527.1K02899WP_250125553.1K02902WP_250122722.1K02904WP_250125985.1K02906WP_250125993.1K02909WP_250125966.1K02911WP_250124624.1K02913WP_250121863.1K02914WP_250123140.1K02916WP_250122605.1K02919WP_250125974.1K02926WP_250125992.1K02931WP_250125982.1K02933WP_250125980.1K02935WP_250125424.1K02939WP_250121540.1K02945WP_250123547.1, WP_250124638.1K02946WP_250121290.1K02948WP_250125972.1K02950WP_250121286.1K02952WP_250125973.1K02954WP_250121552.1K02956WP_250125336.1K02959WP_250124981.1K02961WP_250125984.1K02963WP_250121862.1K02965WP_250125989.1K02967WP_250124818.1K02968WP_250123321.1K02970WP_250122015.1, WP_250122274.1, WP_250122333.1K02982WP_250125987.1K02986WP_250124119.1K02988WP_250125978.1K02990WP_250122951.1K02992WP_250121287.1K02994WP_250125981.1K02996WP_250125967.100970 Aminoacyl-tRNA biosynthesis\\xa0(25)K00604WP_250125215.1K01866WP_250124730.1K01867WP_250126166.1K01868WP_250125326.1K01869WP_250125368.1K01870WP_250123615.1K01872WP_250123557.1K01873WP_250124339.1K01874WP_250123267.1K01875WP_250121555.1K01876WP_250122320.1K01878WP_250121451.1K01879WP_250121991.1K01881WP_250124443.1K01883WP_250121278.1, WP_250126501.1K01885WP_250123393.1K01887WP_250124329.1K01889WP_250123149.1K01890WP_250121868.1K01892WP_250122185.1K01893WP_250124789.1K02433WP_250121584.1, WP_250122150.1K02434WP_250124516.1K02435WP_250122464.1K04567WP_250121498.103008 Ribosome biogenesis in eukaryotes\\xa0(1)K03685WP_250122345.1  Folding, sorting and degradation  03060 Protein export\\xa0(14)K01551WP_250125151.1K03070WP_250123145.1K03072WP_250124360.1K03073WP_250125429.1K03074WP_250124361.1K03075WP_250125958.1K03076WP_250125976.1K03100WP_250121549.1, WP_250122056.1, WP_250124735.1K03101WP_250123095.1K03106WP_250124982.1K03110WP_250125439.1K03116WP_250122331.1, WP_250123883.1K03118WP_250121784.1K03217WP_256478733.104141 Protein processing in endoplasmic reticulum\\xa0(2)K04079WP_250122721.1K13993WP_250125344.1, WP_250126071.1, WP_250126072.104122 Sulfur relay system\\xa0(10)K00566WP_250121598.1K01011WP_250122755.1, WP_250123031.1K03154WP_250121543.1K03635WP_250123857.1K03636WP_250122363.1, WP_250122366.1K03637WP_250122910.1K03638WP_250121271.1K03639WP_250124120.1K04487WP_250125134.1, WP_250125883.1K21147WP_250125482.103018 RNA degradation\\xa0(13)K00937WP_250121710.1K00962WP_250122897.1K01689WP_250125716.1K03654WP_250125213.1, WP_250125324.1K03666WP_250123766.1K04043WP_250122003.1, WP_250122389.1, WP_250122543.1, WP_250126349.1K04077WP_250122604.1, WP_250124781.1K05592WP_250122989.1K08300WP_250121295.1K11927WP_250121888.1K12573WP_250124262.1K12574WP_250124627.1K22468WP_250122842.1  Replication and repair  03030 DNA replication\\xa0(11)K01972WP_250121679.1K02314WP_250121539.1K02316WP_250122123.1K02337WP_250122145.1K02338WP_250125153.1K02340WP_250121983.1K02341WP_250122066.1K02343WP_250123933.1K03111WP_250122517.1K03469WP_250125557.1K03470WP_250121294.103410 Base excision repair\\xa0(10)K01142WP_250126015.1K01247WP_250124583.1K01972WP_250121679.1K03575WP_250124604.1K03649WP_250125234.1K03652WP_250125868.1K07462WP_250121314.1, WP_250125167.1K10563WP_250126148.1K10773WP_250121553.1K21929WP_250122964.103420 Nucleotide excision repair\\xa0(7)K01972WP_250121679.1K03657WP_250124171.1, WP_250125591.1K03701WP_250122461.1K03702WP_250123858.1K03703WP_250126083.1K03723WP_250122433.1K10843WP_250126525.103430 Mismatch repair\\xa0(15)K01972WP_250121679.1K02337WP_250122145.1K02338WP_250125153.1K02340WP_250121983.1K02341WP_250122066.1K02343WP_250123933.1K03111WP_250122517.1K03555WP_250121965.1K03572WP_250123089.1K03601WP_250121492.1K03602WP_250121489.1K03657WP_250124171.1, WP_250125591.1K06223WP_250123176.1K07456WP_250123284.1K07462WP_250121314.1, WP_250125167.103440 Homologous recombination\\xa0(17)K01159WP_250121455.1K02337WP_250122145.1K02338WP_250125153.1K02340WP_250121983.1K02341WP_250122066.1K02343WP_250123933.1K03111WP_250122517.1K03550WP_250123053.1K03551WP_250122501.1K03553WP_250121493.1K03581WP_250121656.1, WP_250126637.1K03584WP_250125952.1K03629WP_250121880.1K03655WP_250124815.1K04066WP_250122904.1K06187WP_250126200.1K07462WP_250121314.1, WP_250125167.1 Environmental Information Processing Membrane transport  02010 ABC transporters\\xa0(89)K01995WP_250124133.1K01996WP_250121449.1K01997WP_250122242.1K01998WP_250124132.1K01999WP_250124166.1K02006WP_250125255.1K02007WP_250125253.1K02008WP_250125254.1K02010WP_250121157.1K02011WP_250121158.1, WP_250122165.1K02012WP_250122542.1K02018WP_250122859.1K02020WP_250122860.1K02036WP_250123093.1, WP_250123396.1K02037WP_250123091.1, WP_250123398.1K02038WP_250123092.1, WP_250123397.1K02040WP_250122421.1, WP_250123074.1, WP_250123090.1, WP_250123399.1, WP_250124113.1, WP_250124114.1K02045WP_250121827.1, WP_250123699.1, WP_250126491.1K02046WP_250124787.1K02047WP_250124786.1K02065WP_250124734.1K02066WP_250126101.1K02067WP_250124733.1K03523WP_250123094.1K05845WP_250125189.1K05846WP_250122140.1K05847WP_250122139.1K06160WP_250121614.1K06861WP_250122291.1K09690WP_250122998.1, WP_250125032.1, WP_250125060.1, WP_250125082.1, WP_250125300.1K09691WP_250123000.1, WP_250124651.1, WP_250125031.1, WP_250125059.1, WP_250125080.1, WP_250125301.1K09815WP_250125474.1K09816WP_250125475.1K09817WP_250126190.1K09969WP_250121977.1K09970WP_250121976.1K09971WP_250121975.1K09972WP_250126225.1K10112WP_250124502.1, WP_250124694.1, WP_275983366.1K10236WP_250124697.1K10237WP_250124696.1K10238WP_250124695.1K10440WP_250124562.1K10820WP_250124563.1K10829WP_250124555.1K11004WP_250122283.1, WP_250123361.1K11069WP_250123648.1K11070WP_250124276.1K11071WP_250123649.1K11072WP_250123647.1K11085WP_250121860.1, WP_250122637.1, WP_250125085.1, WP_250126168.1K11720WP_250122292.1K11950WP_250121845.1K11951WP_250121844.1K11952WP_250121843.1, WP_250122847.1K11953WP_250121842.1K11954WP_250122546.1K11955WP_250121988.1K11956WP_250125405.1K11957WP_250121987.1K11958WP_250125484.1K11959WP_250125386.1K11960WP_250125387.1K11961WP_250125388.1K11962WP_250125389.1K11963WP_250125391.1K14698WP_250126440.1K14699WP_250126446.1K15576WP_250122849.1, WP_250123888.1K15577WP_250122848.1, WP_250123887.1K15578WP_250123889.1K15579WP_250122846.1K16785WP_250125579.1K16786WP_250126053.1K16787WP_250121331.1K16915WP_250125252.1K17062WP_250124274.1K17063WP_250124275.1K17244WP_250122258.1K17245WP_250123326.1K17246WP_250126134.1K18233WP_250123152.1K18889WP_250126121.1K18890WP_250125106.1K23163WP_250124788.1K23227WP_250124556.1K23228WP_250124557.1K23508WP_250124564.1K23509WP_250124561.1': 2,\n         '01057 Biosynthesis of type II polyketide products': 2,\n         '02040 Flagellar assembly': 2,\n         '00311 Penicillin and cephalosporin biosynthesis': 2,\n         '01052 Type I polyketide structures': 2,\n         '00515 Mannose type O-glycan biosynthesis': 1,\n         '00514 Other types of O-glycan biosynthesis': 1,\n         '01040 Biosynthesis of unsaturated fatty acids': 1,\n         '04137 Mitophagy - animal': 1,\n         '02060 Phosphotransferase system': 1,\n         '00981 Insect hormone biosynthesis': 1,\n         '00903 Limonene degradation': 1,\n         '04152 AMPK signaling pathway': 1,\n         '04113 Meiosis - yeast': 1,\n         '00600 Sphingolipid metabolism': 1,\n         '00830 Retinol metabolism': 1,\n         '00626 Naphthalene degradation': 1,\n         '00592 alpha-Linolenic acid metabolism': 1,\n         '00642 Ethylbenzene degradation': 1,\n         '03008 Ribosome biogenesis in eukaryotes': 1,\n         '04151 PI3K-Akt signaling pathway': 1,\n         '04072 Phospholipase D signaling pathway': 1,\n         '03022 Basal transcription factors': 1,\n         '00901 Indole alkaloid biosynthesis': 1,\n         '00253 Tetracycline biosynthesis': 1,\n         '00404 Staurosporine biosynthesis': 1})"
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kos_background"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-06T17:58:26.250874Z",
     "start_time": "2024-04-06T17:58:26.246502Z"
    }
   },
   "id": "408b429c10cf20ba",
   "execution_count": 9
  },
  {
   "cell_type": "code",
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false
   },
   "id": "721968486bc61175"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
